pkg/didcomm/protocol/introduce/service.go

```diff
@@ -348,43 +305,76 @@ func (s *Service) sendMsgEvents(msg *service.StateMsg) {
 }
 
 // newDIDCommActionMsg creates new DIDCommAction message
-func (s *Service) newDIDCommActionMsg(msg *metaData) service.DIDCommAction {
+func (s *Service) newDIDCommActionMsg(md *metaData) service.DIDCommAction {
 	// create the message for the channel
 	// trigger the registered action event
 	actionStop := func(err error) {
 		// if introducee received Proposal disapprove must be true
-		if msg.Msg.Type() == ProposalMsgType {
-			msg.disapprove = true
+		if md.Msg.Type() == ProposalMsgType {
+			md.disapprove = true
 		}
 
-		msg.err = err
-		s.processCallback(msg)
+		md.err = err
+		s.processCallback(md)
 	}
 
 	return service.DIDCommAction{
 		ProtocolName: Introduce,
-		Message:      msg.copyMsg,
-		Continue: func(args interface{}) {
-			// there is no way to receive another interface
-			if dep, ok := args.(InvitationEnvelope); ok {
-				msg.dependency = dep
-				s.processCallback(msg)
-				return
+		Message:      md.msgClone,
+		Continue: func(opt interface{}) {
+			if fn, ok := opt.(Opt); ok {
+				fn(md.Msg.Metadata())
 			}
-			// sets an error to the message
-			actionStop(errMissingDependency)
+
+			if md.Msg.Type() == RequestMsgType {
+				if md.Msg.Metadata()[metaRecipients] == nil {
+					md.err = errors.New("no recipients")
+				}
+			}
+
+			s.processCallback(md)
 		},
-		Stop: func(err error) { actionStop(&customError{error: err}) },
+		Stop: func(err error) { actionStop(customError{error: err}) },
 	}
 }
 
+// Continue allows proceeding with the action by the context ID
+func (s *Service) Continue(ctxID string, opt Opt) error {
```

Comment:
How can the client get the ctxID in case of async? Do we have any API where we can see all the introduce message which require the user/client action ?

Reply:
No, we do not have any API. ctxID for actions is equal to the threadID. Also functions HandleOutbound and HandleInbound will return the context-id.

Reply:
I feel the consumer/clients shouldn't be aware of protocol layer elements such as threadID. They should work with application ids (in case of DIDExchange it's connectionID).

Reply:
And we would need to provide some APIs to lookup the introduce messages by id or state.

Reply:
@rolsonquadras  Why do we need to lookup the introduce messages by id or state?